kafka-broker/src/main/scala/com/lightbend/lagom/internal/broker/kafka/Producer.scala

```diff
@@ -0,0 +1,163 @@
+import scala.concurrent.Future
+import scala.util.Failure
+import scala.util.Success
+
+import org.apache.kafka.clients.producer.ProducerRecord
+import org.apache.kafka.common.serialization.StringSerializer
+import org.slf4j.LoggerFactory
+
+import com.lightbend.lagom.internal.broker.kafka.OffsetTracker.OffsetDao
+import com.lightbend.lagom.internal.broker.kafka.serializer.KafkaSerializer
+import com.lightbend.lagom.javadsl.api.Descriptor.TopicCall
+import com.lightbend.lagom.javadsl.api.broker.modules.Properties.Property
+import com.lightbend.lagom.javadsl.broker.kafka.Properties.PARTITION_KEY_STRATEGY
+import com.lightbend.lagom.javadsl.broker.kafka.property.PartitionKeyStrategy
+import com.lightbend.lagom.javadsl.persistence.Offset
+
+import akka.Done
+import akka.NotUsed
+import akka.actor.Actor
+import akka.actor.ActorLogging
+import akka.actor.ActorSystem
+import akka.actor.PoisonPill
+import akka.actor.Props
+import akka.actor.Status
+import akka.actor.SupervisorStrategy
+import akka.cluster.singleton.ClusterSingletonManager
+import akka.cluster.singleton.ClusterSingletonManagerSettings
+import akka.japi.Pair
+import akka.kafka.ProducerMessage
+import akka.kafka.ProducerSettings
+import akka.kafka.scaladsl.{ Producer => ReactiveProducer }
+import akka.pattern.BackoffSupervisor
+import akka.pattern.pipe
+import akka.stream.FlowShape
+import akka.stream.KillSwitch
+import akka.stream.KillSwitches
+import akka.stream.Materializer
+import akka.stream.javadsl.{ Flow => JFlow }
+import akka.stream.javadsl.{ Source => JSource }
+import akka.stream.scaladsl.Flow
+import akka.stream.scaladsl.GraphDSL
+import akka.stream.scaladsl.Keep
+import akka.stream.scaladsl.Sink
+import akka.stream.scaladsl.Unzip
+import akka.stream.scaladsl.Zip
+
+/**
+ * A Producer for publishing messages in Kafka using the akka-stream-kafka API.
+ */
+class Producer[Message] private (config: KafkaConfig, topicCall: TopicCall[Message], offsetDao: Future[OffsetDao], system: ActorSystem)(implicit mat: Materializer, ec: ExecutionContext) {
+
+  private val log = LoggerFactory.getLogger(classOf[Producer[_]])
+
+  private val producerConfig = ProducerConfig(system.settings.config)
+
+  // Note: The publish method doesn't return a CompletionStage[Done] because this method is not part of the public user api.
+  def publish(readSideStream: JFunction[Offset, JSource[Pair[Message, Offset], NotUsed]]): Unit = {
+    offsetDao.onComplete {
+      case Success(offsetDao) =>
+        val publisherProps = Producer.ProducerActor.props(config, topicCall, readSideStream, offsetDao)
+        val backoffPublisherProps = BackoffSupervisor.propsWithSupervisorStrategy(
+          publisherProps, "kafka-producer", producerConfig.minBackoff, producerConfig.maxBackoff, producerConfig.randomBackoffFactor, SupervisorStrategy.stoppingStrategy
+        )
+
+        val publisherSingletonProps = ClusterSingletonManager.props(backoffPublisherProps, PoisonPill, ClusterSingletonManagerSettings(system))
```

Comment:
No need to use a cluster singleton, as Kafka grouping mechanism will already ensure that only one consumer instance is "active" per group. Hence, I'll remove the ClusterSingletonManager actor.